From d38e13968b79d5069abce10133280426734e26d8 Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Fri, 6 Aug 2021 11:28:06 +0300
Subject: [PATCH] Merge pull request #3404 in SW_UGW/linux from
 DRVLIB_SW-4182-multi-die-spi-nand-flash-support-in-linux-kernel to master

* commit '28a2ce4d8a6f863962bd81a196412b5f679a1e84':
  DRVLIB_SW-4182: Add error checking in waitfunc in spinand driver
---
 drivers/mtd/ltq-spinand/ltq_spinand.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/mtd/ltq-spinand/ltq_spinand.c b/drivers/mtd/ltq-spinand/ltq_spinand.c
index 76b9d11de9b5..c61441f5a1bf 100644
--- a/drivers/mtd/ltq-spinand/ltq_spinand.c
+++ b/drivers/mtd/ltq-spinand/ltq_spinand.c
@@ -1181,6 +1181,9 @@ static void spinand_reset(struct spi_device *spi)
 		spinand_set_otp_mask(spi, OTP_ECC_ENABLE, OTP_ECC_ENABLE);
 	}
 
+	/* set die selection back to 0 after reset */
+	info->current_die = 0;
+
 	return;
 }
 
@@ -1306,8 +1309,12 @@ static int spinand_wait(struct mtd_info *mtd, struct nand_chip *chip)
 		retval = spinand_read_status(info->spi, &status);
 		if (retval < 0)
 			break;
-		if ((status & STATUS_OIP_MASK) == STATUS_READY)
-			return 0;
+		if ((status & STATUS_OIP_MASK) == STATUS_READY) {
+			if ((status & STATUS_E_FAIL_MASK) == STATUS_E_FAIL)
+				return NAND_STATUS_FAIL;
+			else
+				return 0;
+		}
 
 		cond_resched();
 	}
